pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/invertedindex/RealtimeNativeTextIndex.java

```diff
@@ -0,0 +1,146 @@
+import org.apache.pinot.segment.spi.index.mutable.MutableTextIndex;
+import org.roaringbitmap.PeekableIntIterator;
+import org.roaringbitmap.RoaringBitmapWriter;
+import org.roaringbitmap.buffer.ImmutableRoaringBitmap;
+import org.roaringbitmap.buffer.MutableRoaringBitmap;
+
+
+public class RealtimeNativeTextIndex implements MutableTextIndex {
+  private final String _column;
+  private final MutableFST _mutableFST;
+  private final RealtimeInvertedIndex _invertedIndex;
+  private final Map<String, Integer> _termToDictIdMapping;
+  private int _nextDocId = 0;
+  private int _nextDictId = 0;
+  private final ReentrantReadWriteLock.ReadLock _readLock;
+  private final ReentrantReadWriteLock.WriteLock _writeLock;
+
+  public RealtimeNativeTextIndex(String column) {
+    _column = column;
+    _mutableFST = new MutableFSTImpl();
+    _termToDictIdMapping = new HashMap<>();
+    _invertedIndex = new RealtimeInvertedIndex();
+
+    ReentrantReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+    _readLock = readWriteLock.readLock();
+    _writeLock = readWriteLock.writeLock();
+  }
+
+  @Override
+  public void add(String document) {
+    List<String> tokens;
+    try {
+      tokens = analyze(document, new StandardAnalyzer(LuceneTextIndexCreator.ENGLISH_STOP_WORDS_SET));
+    } catch (IOException e) {
+      throw new RuntimeException(e.getMessage());
+    }
+
+    try {
+      _writeLock.lock();
+      for (String token : tokens) {
+        Integer currentDictId = _termToDictIdMapping.get(token);
+        if (currentDictId == null) {
+
+          currentDictId = _nextDictId;
+          _mutableFST.addPath(token, currentDictId);
+          _termToDictIdMapping.put(token, currentDictId);
+          ++_nextDictId;
+        }
+        addToPostingList(currentDictId);
+      }
+      _nextDocId++;
+    } finally {
+      _writeLock.unlock();
+    }
+  }
+
+  @Override
+  public ImmutableRoaringBitmap getDictIds(String searchQuery) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public MutableRoaringBitmap getDocIds(String searchQuery) {
+    MutableRoaringBitmap matchingDocIds = null;
+    try {
+      _readLock.lock();
+      RoaringBitmapWriter<MutableRoaringBitmap> writer = RoaringBitmapWriter.bufferWriter().get();
+      RealTimeRegexpMatcher.regexMatch(searchQuery, _mutableFST, writer::add);
+      ImmutableRoaringBitmap matchingDictIds = writer.get();
+
+      for (PeekableIntIterator it = matchingDictIds.getIntIterator(); it.hasNext(); ) {
+        int dictId = it.next();
+        if (dictId >= 0) {
+          ImmutableRoaringBitmap docIds = _invertedIndex.getDocIds(dictId);
+          if (matchingDocIds == null) {
+            matchingDocIds = docIds.toMutableRoaringBitmap();
+          } else {
+            matchingDocIds.or(docIds);
+          }
+        }
+      }
+    } finally {
+      _readLock.unlock();
+    }
+
+    return matchingDocIds == null ? new MutableRoaringBitmap() : matchingDocIds;
```

Comment:
No need to create a bitmap for the dict ids. This can be simplified
```suggestion
    MutableRoaringBitmap matchingDocIds = new MutableRoaringBitmap();
    _readLock.lock();
    try {
      RealTimeRegexpMatcher.regexMatch(searchQuery, _mutableFST, dictId -> matchingDocIds.or(_invertedIndex.getDocIds(dictId)));
      return matchingDocIds;
      }
    } finally {
      _readLock.unlock();
    }
```